### ISSUE TITLE
eeshenggoh - distribute() cannot be called causing denial of service bricking the protocol and lesser rewards emissions for veNFT holders

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-06-velocimeter-judging#244]
eeshenggoh

High

# distribute() cannot be called causing denial of service bricking the protocol and lesser rewards emissions for veNFT holders

## Summary
From docs:
> In V4, this is flipped. Emissions are now a factor of how many gauges there are. The more gauges there are the more emissions are minted. That being said, voting still fully directs emissions. If a gauge gets 5% of the total votes it will only still get 5% of the emissions even though the addition of that gauge may increase emissions by more than 5%. 
https://paragraph.xyz/@velocimeter/velocimeter-v4

The proxy gauge, created by the development team and distinct from the Gaugev4 type, the team uses ProxyGauges specifically for FLOW/WETH LP token. 

From Sherlock README.md:
> Are there any off-chain mechanisms or off-chain procedures for the protocol (keeper bots, arbitrage bots, etc.)?
> Function distribute on voter contract needs to be called every epoch ( one week), it can be called by anybody but team have keepers setup that call it as well.

We know that distribute will be called, but the problem is if proxyGauge address is included, the function will result in a denial of service.

## Vulnerability Detail

The denial of service comes in when anyone or keeper bot the `distribute()` and in turn calls `distribute(address _gauge)` which will call FLOW/WETH proxy gauge via `proxyGauge::notifyRewardAmount`.
```solidity
// Voter.sol
    function distribute(address _gauge) public lock {
        IMinter(minter).update_period();
        _updateFor(_gauge); 
        uint _claimable = claimable[_gauge];
        if (_claimable > IGauge(_gauge).left(base) && _claimable / DURATION > 0) {
            claimable[_gauge] = 0;
            if((_claimable * 1e18) / currentEpochRewardAmount > minShareForActiveGauge) {
                activeGaugeNumber += 1;
            }

            IGauge(_gauge).notifyRewardAmount(base, _claimable); //@audit
            emit DistributeReward(msg.sender, _gauge, _claimable);
        }
    }
```

From the Sherlock README.md:
> Are there any limitations on values set by admins (or other roles) in the codebase, including restrictions on array lengths?
No
Example deployment script with the default start settings is available in our repository in script folder.
It is used as a template for our deployments.
It has two parts - first Deployment.s.sol script needs to be run then `OFlowDeployment.s.sol` script.

Following the standard of how contracts are deployed, in` Deployment.s.sol`, the  _pairfactory and proxyGaugeFactory(flow)_ were added into `Voter.sol`. This `Voter::addFactory` function allows the updating of proxyGauges to be used in `Voter::creategauge`.
```solidity
        voter.addFactory( address(pairFactory), address(proxyFactory));
```
In `OFlowDeployment.s.sol`, the team deploys the proxy gauge and calls the `Voter::createGauge` to update the internal accounting.
```solidity
        address pair = PairFactory(NEW_PAIR_FACTORY).getPair(
            NEW_FLOW,
            WETH,
            false
        );
        -- SNIP --
        ProxyGaugeFactory(NEW_PROXY_GAUGE_FACTORY).deployGauge(NEW_REWARDS_DISTRIBUTOR_FLOW,pair, "veNFT");
        -- SNIP --
        Voter(NEW_VOTER).createGauge(pair, 1); //why create for FLOW/WETH? @audit
```

This function will deploy the new ProxyGauge for FLOW/WETH LP.
```solidity
    function deployGauge(address _notifyAddress,address _pairAddress,string memory _symbol) external onlyOwner returns (address) {
        address last_gauge = address(new ProxyGauge(flow,_notifyAddress,_symbol)); //@audit
        if(_pairAddress == address(0)) {
            _pairAddress = last_gauge;
        }
        isWhitelisted[_pairAddress] = last_gauge;
        return last_gauge;
    } 
```
Here is the ProxyGauge contract:
```solidity
// ProxyGauge.sol
    constructor(address _flow,address _notifyAddress,string memory _symbol) {
        flow = _flow;
        notifyAddress = _notifyAddress;
        symbol = _symbol;
    }

    function notifyRewardAmount(address token, uint amount) external lock {
        require(token == flow);
        require(amount > 0);
        
        uint256 balanceBefore = IERC20(token).balanceOf(address(this));
        _safeTransferFrom(token, msg.sender, address(this), amount);
        uint256 balanceAfter = IERC20(token).balanceOf(address(this));

        amount = balanceAfter - balanceBefore;

        _safeApprove(flow, notifyAddress, amount);
        IProxyGaugeNotify(notifyAddress).notifyRewardAmount(amount); //@audit doesn't work
    }
```
The `distribute()` function will iterate all gauges types.  For FLOW/WETH proxy gauges the `ProxyGauge ::notifyRewardAmount` will be called. However, there is no `notifyRewardAmount` in the `Pair.sol`. Hence when you call a function on an address that does not implement that function, the call will revert. This means that the transaction will fail and any state changes made before this call will be rolled back unable to update the activeGaugeNumber for FLOW/WETH gauge.

Another point is that the `distribute()` function is the most important because it affects the weekly emissions of tokens. This means that the number of emissions rewards will be lesser if the FLOW/WETH proxy gauge contract can't be called.
```solidity
    function weekly_emission() public view returns (uint) {
        uint256 numberOfGauges = _voter.activeGaugeNumber(); //@audit 
        if(numberOfGauges == 0) { 
            return weeklyPerGauge;
        }
        return weeklyPerGauge * numberOfGauges;
    }
```
## Impact
1) Anyone who calls `distribute` using ProxyGauge address will fail, in this case anyone including the keeper bot will always fail to call `distribute()`, causing a denial of service. 
2) The `activeGaugeNumber` used for factorization with weeklyEmission cannot be included, causing lesser rewards.

## Code Snippet
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/Voter.sol#L549C1-L562C6
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/ProxyGauge.sol#L45

## Tool used

Manual Review

## Recommendation
Include `notifyRewardAmount` logic in pair contract to prevent DoS and accurate weekly reward emission.

Duplicate of #122

**Comment by nevillehuang** [comment]
invalid, deployment scripts are OOS, so the appropriate `notifyAddress` is expected to be inputted

**Comment by goheesheng** [comment]
Hi @nevillehuang, I believe that it is not a duplicate of #122. My report is more of a DoS than math calculation. Isn't README.md the highest hierarchy of truth, hence when I submitted this issue, I followed the standard from how it was deployed:
>Are there any limitations on values set by admins (or other roles) in the codebase, including restrictions on array lengths?
No
Example deployment script with the default start settings is available in our repository in script folder.
It is used as a template for our deployments.
It has two parts - first Deployment.s.sol script needs to be run then OFlowDeployment.s.sol script.

Based on what's written in the README, the current Sherlock Rules don't explicitly define the hierarchy between OOS files and the README. Therefore, I believe that since `Pair::notifyRewardAmount` is not present in the `Pair.sol` contract, it can't be called, which could result in a DoS (Denial of Service).
